In [1]:
import pandas as pd
import numpy as np
import csv
import matplotlib.pyplot as plt
plt.style.use('ggplot')
pd.set_option('display.max_columns', 200)

from plotly.subplots import make_subplots
import plotly.express as px

import warnings
warnings.filterwarnings(action='ignore')

In [2]:
df =  pd.read_csv('../data/Dataset.csv') # собранные нами данные 

In [3]:
df

SEX  How many times do you eat a day? How many hours a day do you sleep?  \
0      M                                 1                                2-4   
1      F                                 3                                5-6   
2      F                                 3                                5-6   
3      F                                 3                               9-10   
4      F                                 4                                7-8   
...   ..                               ...                                ...   
1176   M                                 2                                2-4   
1177   M                                 4                                2-4   
1178   M                                 5                               9-10   
1179   F                                 3                               More   
1180   F                                 5                               More   

     How many times per week do you eat fastfood?  \
0                                               0   
1                                               1   
2                                               0   
3                                               3   
4                                               1   
...                                           ...   
1176                                            0   
1177                                            1   
1178                                            2   
1179                                            2   
1180                                            0   

     Average number of steps per day? Cow's milk  \
0                              0-4000        Yes   
1                              0-4000        Yes   
2                          7000-10000        Yes   
3                           4000-7000         No   
4                           4000-7000  Sometimes   
...                               ...        ...   
1176                           15000+  Sometimes   
1177                        4000-7000         No   
1178                       7000-10000  Sometimes   
1179                           15000+  Sometimes   
1180                      10000-15000         No   

     Vegetable drinks (eg soy milk) Low-fat white yogurt Fresh cheeses  \
0                               Yes                  Yes           Yes   
1                                No                  Yes           Yes   
2                               Yes                   No     Sometimes   
3                               Yes                   No           Yes   
4                               Yes            Sometimes           Yes   
...                             ...                  ...           ...   
1176                      Sometimes                  Yes            No   
1177                             No            Sometimes     Sometimes   
1178                             No            Sometimes     Sometimes   
1179                      Sometimes            Sometimes           Yes   
1180                             No                   No     Sometimes   

           Meat Processed Meat (es prosciutto)       Fish       Eggs  \
0           Yes                            Yes        Yes        Yes   
1           Yes                             No        Yes        Yes   
2           Yes                      Sometimes        Yes        Yes   
3           Yes                             No        Yes        Yes   
4            No                             No  Sometimes        Yes   
...         ...                            ...        ...        ...   
1176  Sometimes                            Yes         No        Yes   
1177         No                      Sometimes        Yes  Sometimes   
1178        Yes                            Yes        Yes         No   
1179        Yes                            Yes  Sometimes  Sometimes   
1180  Sometimes                             No         No         No   

        Legumes 

### Нахождение и удаление строк-дубликатов

In [4]:
df = df.drop_duplicates()
df

SEX  How many times do you eat a day? How many hours a day do you sleep?  \
0      M                                 1                                2-4   
1      F                                 3                                5-6   
2      F                                 3                                5-6   
3      F                                 3                               9-10   
4      F                                 4                                7-8   
...   ..                               ...                                ...   
1176   M                                 2                                2-4   
1177   M                                 4                                2-4   
1178   M                                 5                               9-10   
1179   F                                 3                               More   
1180   F                                 5                               More   

     How many times per week do you eat fastfood?  \
0                                               0   
1                                               1   
2                                               0   
3                                               3   
4                                               1   
...                                           ...   
1176                                            0   
1177                                            1   
1178                                            2   
1179                                            2   
1180                                            0   

     Average number of steps per day? Cow's milk  \
0                              0-4000        Yes   
1                              0-4000        Yes   
2                          7000-10000        Yes   
3                           4000-7000         No   
4                           4000-7000  Sometimes   
...                               ...        ...   
1176                           15000+  Sometimes   
1177                        4000-7000         No   
1178                       7000-10000  Sometimes   
1179                           15000+  Sometimes   
1180                      10000-15000         No   

     Vegetable drinks (eg soy milk) Low-fat white yogurt Fresh cheeses  \
0                               Yes                  Yes           Yes   
1                                No                  Yes           Yes   
2                               Yes                   No     Sometimes   
3                               Yes                   No           Yes   
4                               Yes            Sometimes           Yes   
...                             ...                  ...           ...   
1176                      Sometimes                  Yes            No   
1177                             No            Sometimes     Sometimes   
1178                             No            Sometimes     Sometimes   
1179                      Sometimes            Sometimes           Yes   
1180                             No                   No     Sometimes   

           Meat Processed Meat (es prosciutto)       Fish       Eggs  \
0           Yes                            Yes        Yes        Yes   
1           Yes                             No        Yes        Yes   
2           Yes                      Sometimes        Yes        Yes   
3           Yes                             No        Yes        Yes   
4            No                             No  Sometimes        Yes   
...         ...                            ...        ...        ...   
1176  Sometimes                            Yes         No        Yes   
1177         No                      Sometimes        Yes  Sometimes   
1178        Yes                            Yes        Yes         No   
1179        Yes                            Yes  Sometimes  Sometimes   
1180  Sometimes                             No         No         No   

        Legumes 

In [5]:
df.isna().sum()

SEX                                                                                                           0
How many times do you eat a day?                                                                              0
How many hours a day do you sleep?                                                                            0
How many times per week do you eat fastfood?                                                                  0
Average number of steps per day?                                                                              0
Cow's milk                                                                                                    0
Vegetable drinks (eg soy milk)                                                                                0
Low-fat white yogurt                                                                                          0
Fresh cheeses                                                                                           

### Посмотрим, какие уникальные ответы есть в каждой колонке, и преобразуем их


In [6]:
for column in df.columns:
    unique_values = df[column].unique()
    print(len(unique_values))
    print(f"Unique values in '{column}': {unique_values}")

2
Unique values in 'SEX': ['M' 'F']
7
Unique values in 'How many times do you eat a day?': [1 3 4 2 5 6 7]
5
Unique values in 'How many hours a day do you sleep?': ['2-4' '5-6' '9-10' '7-8' 'More']
8
Unique values in 'How many times per week do you eat fastfood?': ['0' '1' '3' '2' '5' '4' '6' '7+']
5
Unique values in 'Average number of steps per day?': ['0-4000' '7000-10000' '4000-7000' '10000-15000' '15000+']
4
Unique values in 'Cow's milk': ['Yes' 'No' 'Sometimes' "Don't know"]
4
Unique values in 'Vegetable drinks (eg soy milk)': ['Yes' 'No' 'Sometimes' "Don't know"]
4
Unique values in 'Low-fat white yogurt': ['Yes' 'No' 'Sometimes' "Don't know"]
4
Unique values in 'Fresh cheeses': ['Yes' 'Sometimes' 'No' "Don't know"]
3
Unique values in 'Meat': ['Yes' 'No' 'Sometimes']
4
Unique values in 'Processed Meat (es prosciutto)': ['Yes' 'No' 'Sometimes' "Don't know"]
3
Unique values in 'Fish': ['Yes' 'Sometimes' 'No']
4
Unique values in 'Eggs': ['Yes' 'Sometimes' 'No' "Don't know"]
4
Unique 

## Что нужно поменять?
Объединить ответы в столбцах
- 'Do you ever miss meals?
- 'When are you hungry during the day?
- 'How many times do you consume alcoholic beverages in a week?'

### Преобразование значений столбца 'When are you hungry during the day?

In [7]:
df['When are you hungry during the day?'].unique()

array(['When I wake Up', 'During the morning,In the afternoon',
       'After lunch', 'During the morning,Before lunch,Before Dinner',
       'In the afternoon,Before Dinner', 'When I wake Up,After lunch',
       'Before lunch,After Dinner',
       "Before Dinner,After Dinner,I'm always hungry",
       'When I wake Up,Before lunch,Before Dinner',
       'In the afternoon,After Dinner', 'Before lunch,Before Dinner',
       'During the morning,In the afternoon,Before Dinner',
       'When I wake Up,During the morning,In the afternoon,Before lunch,Before Dinner',
       'When I wake Up,Before Dinner', 'After lunch,Before Dinner',
       'Before Dinner', 'During the morning,Before Dinner',
       'In the afternoon', "I'm always hungry", 'During the morning',
       'Before lunch',
       'When I wake Up,During the morning,Before lunch,Before Dinner',
       'When I wake Up,In the afternoon,Before Dinner', 'After Dinner',
       'During the morning,After Dinner', 'In the afternoon,Before lu

In [8]:
len(df['When are you hungry during the day?'].unique())

78

***Финальные уникальные значения***
1. Утром - in the morning
2. Во второй половине дня - in the afternoon
3. Вечером - in the evening
4. В первой и второй половине дня - morning+afternoon
5. Во второй и третьей - afternoon+evening
6. В третьей и первой - morning+evening
7. Я всегда голоден  - I am always hungry

In [9]:
df['When are you hungry during the day?'] = df['When are you hungry during the day?'].replace(['When I wake Up,During the morning', 'When I wake Up', 'During the morning'], 'In the morning')
df['When are you hungry during the day?'] = df['When are you hungry during the day?'].replace(['After lunch', 'Before lunch,After lunch', 'In The Afternoon', 'In the afternoon,After lunch', 'In the afternoon,Before lunch', 'In the afternoon', 'Before lunch'], 'In the afternoon')
df['When are you hungry during the day?'] = df['When are you hungry during the day?'].replace(['Before Dinner,After Dinner', 'Before Dinner', 'After Dinner'], 'In the evening')
df['When are you hungry during the day?'] = df['When are you hungry during the day?'].replace(['During the morning,In the afternoon', 'During the morning,In the afternoon,Before lunch', 'During the morning,Before lunch', 'During the morning,After lunch', 'When I wake Up,Before lunch', 'When I wake Up,During the morning,Before lunch', 'When I wake Up,After lunch'], 'morning+afternoon')
df['When are you hungry during the day?'] = df['When are you hungry during the day?'].replace(['Before lunch,After lunch,Before Dinner,After Dinner', 'In the afternoon,After lunch,Before Dinner', 'In the afternoon,Before lunch,After Dinner', 'After lunch,After Dinner', 'Before lunch,Before Dinner,After Dinner', 'In the afternoon,After lunch,After Dinner', 'In the afternoon,Before lunch,Before Dinner,After Dinner', 'In the afternoon,Before lunch,After lunch,Before Dinner,After Dinner', 'In the afternoon,Before lunch,After lunch,Before Dinner', 'In the afternoon,Before lunch,Before Dinner', 'After lunch,Before Dinner,After Dinner', 'In the afternoon,Before Dinner,After Dinner', 'After lunch,Before Dinner', 'In the afternoon,After Dinner', 'Before lunch,Before Dinner', 'Before lunch,After Dinner', 'In the afternoon,Before Dinner'], 'afternoon+evening')
df['When are you hungry during the day?'] = df['When are you hungry during the day?'].replace([ 'When I wake Up,During the morning,After Dinner','During the morning,Before Dinner,After Dinner' ,'When I wake Up,During the morning,Before Dinner','When I wake Up,After Dinner','During the morning,After Dinner', 'During the morning,Before Dinner', 'When I wake Up,Before Dinner'],'morning+evening')
len(df['When are you hungry during the day?'].unique())

41

In [10]:
 df['When are you hungry during the day?'] = df['When are you hungry during the day?'].replace(['During the morning,Before lunch,Before Dinner',
    'During the morning,In the afternoon,Before Dinner',
   'When I wake Up,During the morning,In the afternoon,Before lunch,Before Dinner',
   'When I wake Up,During the morning,In the afternoon,Before lunch,After lunch,Before Dinner,After Dinner',
    "During the morning,Before lunch,I'm always hungry",
   'During the morning,In the afternoon,Before lunch,Before Dinner',
     'When I wake Up,Before lunch,Before Dinner,After Dinner',
   "I'm always hungry",
       'During the morning,After lunch,After Dinner',
   "In the afternoon,Before lunch,Before Dinner,I'm always hungry",
       'During the morning,Before lunch,After Dinner',
       'When I wake Up,After lunch,After Dinner',
       "In the afternoon,Before Dinner,After Dinner,I'm always hungry",
       "When I wake Up,During the morning,In the afternoon,Before lunch,Before Dinner,I'm always hungry",
       "During the morning,In the afternoon,I'm always hungry",
 "After lunch,I'm always hungry",
       'During the morning,After lunch,Before Dinner',
       "Before lunch,Before Dinner,I'm always hungry",
       "When I wake Up,Before lunch,Before Dinner,I'm always hungry",
       "When I wake Up,During the morning,In the afternoon,Before lunch,After lunch,Before Dinner,After Dinner,I'm always hungry",
       'When I wake Up,Before lunch,After Dinner',
       'During the morning,In the afternoon,After Dinner',
       'During the morning,In the afternoon,After lunch,Before Dinner,After Dinner',
       'When I wake Up,After lunch,Before Dinner',
     "Before lunch,Before Dinner,After Dinner,I'm always hungry",
    "In the afternoon,I'm always hungry",
  'When I wake Up,During the morning,In the afternoon,Before Dinner',
       'When I wake Up,During the morning,After lunch,Before Dinner',
     "When I wake Up,I'm always hungry",
  "When I wake Up,In the afternoon,After lunch,After Dinner,I'm always hungry",
       'When I wake Up,In the afternoon,Before lunch,Before Dinner',
  'When I wake Up,During the morning,Before lunch,Before Dinner',
       'When I wake Up,In the afternoon,Before Dinner',
  "Before Dinner,After Dinner,I'm always hungry",
       'When I wake Up,Before lunch,Before Dinner'], "I'm always hungry")
len(df['When are you hungry during the day?'].unique())

7

In [11]:
df['When are you hungry during the day?'].unique()

array(['In the morning', 'morning+afternoon', 'In the afternoon',
       "I'm always hungry", 'afternoon+evening', 'morning+evening',
       'In the evening'], dtype=object)

### Преобразования столбца "Do you ever miss meals?"

***Как хотим поменять?***

'No',

'Yes', 

'Yes, breakfast',

'Yes for craving' -Yes, 

'I have no time' -Yes,

'Salto spesso i pasti perché non ho tempo' -Yes,

'Yes lunch,I have no time' - Yes, lunch, 

'Yes lunch,Yes dinner,I have no time' - Yes,lunch + dinner,

'Yes, breakfast', 

'Yes, lunch', 

'Yes breakfast,I have no time' -Yes breakfast,

'Yes dinner',

"No,I'm always hungry" - No, 

'Yes breakfast,Yes lunch'- Yes, breakfast+lunch,

"I'm always hungry" - No,

'Yes breakfast,Yes lunch,Yes dinner' - Yes,

'Yes breakfast,Yes dinner' - Yes, dinner+breakfast ,

'Yes breakfast,Yes lunch,I have no time' - Yes, breakfast+lunch,

"Yes breakfast,Yes lunch,I have no time,I'm always hungry" - Yes, breakfast+lunch,

"I have no time,I'm always hungry" - Yes,

"Yes breakfast,I have no time,I'm always hungry" - Yes, breakfast,

'Yes breakfast,Yes dinner,I have no time' - Yes, breakfast + dinner,

"Yes dinner,I'm always hungry" - Yes, dinner,

'Yes lunch,Yes dinner' - Yes, lunch+dinner,

"Yes dinner,I have no time' - Yes, dinner, 

'Yes breakfast,I'm always hungry" - Yes, breakfast,

"Yes breakfast,Yes lunch,Yes dinner,I have no time,I'm always hungry" - Yes



***Финальные уникальные значения***

1. No

2. Yes

3. Yes, breakfast

4. Yes, lunch

5. Yes, dinner

6. Yes, breakfast+lunch

7. Yes, lunch+dinner

8. Yes, dinner+breakfast



In [12]:
len(df['Do you ever miss meals?'].unique())

23

In [13]:
def transform_meal_response(value):
    mapping = {
        'Yes for craving': 'Yes',
        'I have no time': 'Yes',
        'Salto spesso i pasti perché non ho tempo': 'Yes',
        'Yes lunch,I have no time': 'Yes, lunch',
        'Yes lunch,Yes dinner,I have no time': 'Yes, lunch+dinner',
        'Yes breakfast,I have no time': 'Yes, breakfast',
        'Yes breakfast': 'Yes, breakfast',
        'Yes dinner': 'Yes, dinner',
        "No,I'm always hungry": 'No',
        'Yes breakfast,Yes lunch': 'Yes, breakfast+lunch',
        "I'm always hungry": 'No',
        'Yes lunch': 'Yes, lunch',
        'Yes breakfast,Yes lunch,Yes dinner': 'Yes',
        'Yes breakfast,Yes dinner': 'Yes, breakfast+dinner',
        'Yes breakfast,Yes lunch,I have no time': 'Yes, breakfast+lunch',
        'Yes breakfast,Yes lunch,I have no time,Я всегда голоден': 'Yes, breakfast+lunch',
        'I have no time,Я всегда голоден': 'Yes',
        'Yes breakfast,I have no time,Я всегда голоден': 'Yes, breakfast',
        'Yes breakfast,Yes dinner,I have no time': 'Yes, breakfast+dinner',
        'Yes dinner,Я всегда голоден': 'Yes, dinner',
        'Yes lunch,Yes dinner': 'Yes, lunch+dinner',
        'Yes dinner,I have no time': 'Yes, dinner',
        'Yes breakfast,Я всегда голоден': 'Yes, breakfast',
        'Yes breakfast,Yes lunch,Yes dinner,I have no time,Я всегда голоден': 'Yes'
    }
    
    return mapping.get(value, value)

In [14]:
df['Do you ever miss meals?'] = df['Do you ever miss meals?'].apply(transform_meal_response)

unique_values = df['Do you ever miss meals?'].unique()
len(unique_values)

14

In [15]:
unique_values

array(['No', 'Yes, lunch', 'Yes, lunch+dinner', 'Yes', 'Yes, breakfast',
       'Yes, dinner', 'Yes, breakfast+lunch', 'Yes, breakfast+dinner',
       "Yes breakfast,Yes lunch,I have no time,I'm always hungry",
       "I have no time,I'm always hungry",
       "Yes breakfast,I have no time,I'm always hungry",
       "Yes dinner,I'm always hungry", "Yes breakfast,I'm always hungry",
       "Yes breakfast,Yes lunch,Yes dinner,I have no time,I'm always hungry"],
      dtype=object)

### Преобразование столбца 'How many times do you consume alcoholic beverages in a week?'

In [16]:
df['How many times do you consume alcoholic beverages in a week?'].unique()

array(['0', '1', '4', '3', '2', '5', '7+', '6'], dtype=object)

In [17]:
# считаем количество ответов '7+'
count_7_plus = df['How many times do you consume alcoholic beverages in a week?'].value_counts()['7+']
count_7_plus

3

количество ответов '7+' очень мало, можем изменить на 7

более того, всю колонку можно перевести в integer

In [18]:
df['How many times do you consume alcoholic beverages in a week?'] = df['How many times do you consume alcoholic beverages in a week?'].replace('7+', 7)

# преобразуем всю колонку в int
df['How many times do you consume alcoholic beverages in a week?'] = df['How many times do you consume alcoholic beverages in a week?'].astype(int)

df['How many times do you consume alcoholic beverages in a week?'].unique()

array([0, 1, 4, 3, 2, 5, 7, 6])

### Потребление фастфуда

In [19]:
df['How many times per week do you eat fastfood?'].unique()

array(['0', '1', '3', '2', '5', '4', '6', '7+'], dtype=object)

In [20]:
count_7_plus =df['How many times per week do you eat fastfood?'].value_counts()['7+']
count_7_plus

5

In [21]:
df['How many times per week do you eat fastfood?'] = df['How many times per week do you eat fastfood?'].replace('7+', 7)


df['How many times per week do you eat fastfood?'] = df['How many times per week do you eat fastfood?'].astype(int)

df['How many times per week do you eat fastfood?'].unique()

array([0, 1, 3, 2, 5, 4, 6, 7])

## Целевая переменная

In [22]:
df['Average number of steps per day?'].unique()

array(['0-4000', '7000-10000', '4000-7000', '10000-15000', '15000+'],
      dtype=object)

In [23]:
def categorize_steps(num):
    if num == '0-4000':
        return 0
    elif num == '4000-7000':
        return 1
    elif num == '7000-10000':
        return 2
    elif num == '10000-15000':
        return 3
    else:
        return 4
        
df['Average number of steps per day?'] = df['Average number of steps per day?'].apply(categorize_steps)

df['Average number of steps per day?'].unique()

array([0, 2, 1, 3, 4])

# Статистический анализ

In [24]:
columns = df.columns.values

columns = np.delete(columns, np.where(columns == 'Average number of steps per day?'))

#### Из блокнота про сон знаем, для каких колонок нужно провести провести корреляционный анализ Спирмена

In [25]:
from scipy.stats import spearmanr

In [26]:
questions_spearman = [
    'How many times do you eat a day?',
    'You like salty or sweet? (where 10 - "I love salt and hate sweet" and 1 - "I am the biggest fan of sweet")',
    'How many liters of water do you drink per day?',
    'How many times do you consume alcoholic beverages in a week?',
    'Age',
    'How many sugary drinks do you consume per day?',
    'How many times per week do you eat fastfood?'
]


In [27]:
for column in questions_spearman:
    print(column, df[column].unique())

How many times do you eat a day? [1 3 4 2 5 6 7]
You like salty or sweet? (where 10 - "I love salt and hate sweet" and 1 - "I am the biggest fan of sweet") [ 1  4  8  5  6  7  2  3 10  9]
How many liters of water do you drink per day? [0 1 3 2 4 5]
How many times do you consume alcoholic beverages in a week? [0 1 4 3 2 5 7 6]
Age ['0-14' '15-20' '21-30' '31-50' '51-65' '65+']
How many sugary drinks do you consume per day? [0 2 5 1 4 3 7 6]
How many times per week do you eat fastfood? [0 1 3 2 5 4 6 7]


In [28]:
for column in questions_spearman:
    correlation, p_value = spearmanr(df[column], df['Average number of steps per day?'])
    print(f'{column} correlation - {correlation}, p_value - {p_value}\n====================')

How many times do you eat a day? correlation - 0.07949305104831342, p_value - 0.00631528264869372
You like salty or sweet? (where 10 - "I love salt and hate sweet" and 1 - "I am the biggest fan of sweet") correlation - 0.11682695963009511, p_value - 5.796194661993929e-05
How many liters of water do you drink per day? correlation - 0.20893344276440556, p_value - 4.2746261722872855e-13
How many times do you consume alcoholic beverages in a week? correlation - 0.006093650997159617, p_value - 0.834436476309126
Age correlation - 0.06505189457495988, p_value - 0.025505734244528806
How many sugary drinks do you consume per day? correlation - 0.055883516290439264, p_value - 0.05507145911883581
How many times per week do you eat fastfood? correlation - -0.0052057929205309664, p_value - 0.8582826757705304


1. How many times do you eat a day?  
   - Коэффициент корреляции: 0.079  
   - p-value: 0.006 > 0.005
   
   Вывод: очень слабая положительная корреляция между количеством приемов пищи в день и количеством шагов

2. You like salty or sweet? (where 10 - "I love salt and hate sweet" and 1 - "I am the biggest fan of sweet")  
   - Коэффициент корреляции: 0.117  
   - p-value: 5.796e-05 < 0.005
   
   Вывод:  положительная корреляция между предпочтением соленого или сладкого и количеством шагов

3. How many liters of water do you drink per day?  
   - Коэффициент корреляции: 0.209  
   - p-value: 4.275e-13 < 0.005
   
   Вывод: положительная корреляция между потреблением литров воды в день и количеством шагов

4. How many times do you consume alcoholic beverages in a week?  
   - Коэффициент корреляции: 0.006  
   - p-value: 0.834  > 0.005
   
   Вывод: Нет статистически значимой корреляции между употреблением алкоголя в неделю и количеством шагов

5. Age  
   - Коэффициент корреляции: 0.065  
   - p-value: 0.026  < 0.005
   
   Вывод: Наблюдается слабая положительная корреляция между возрастом и количеством шагов

6. How many sugary drinks do you consume per day?  
   - Коэффициент корреляции: 0.056  
   - p-value: 0.055 > 0.005 
   
   Вывод: очень слабая положительная корреляция между потреблением сладких напитков и количеством шагов

7. How many times per week do you eat fastfood?
    - Коэффициент корреляции: -0.005
    - p-value: 0.858
   Вывод:   корреляция близка к нулю свидетельствует о том, что нет статистически значимой связи между уровнем потребления фастфуда в неделю и количеством шагов

#### Теперь хи-квадрат тест

In [29]:
from scipy.stats import chi2_contingency

In [30]:
questions_chi_square = list(set(columns) - set(questions_spearman))


#### Преобразование колонок

#### Качество сна

In [31]:
def categorize_sleep(quality):
    if quality in ['I struggle to fall asleep', 'I wake up several times during the night', 'I wake up a lot sooner than I would like', 'I struggle to fall asleep,I wake up several times during the night',
                   'Good,I struggle to fall asleep','I struggle to fall asleep,I wake up several times during the night,I wake up a lot sooner than I would like',
                  'I wake up several times during the night,I wake up a lot sooner than I would like','Good,I wake up several times during the night',
                   'Good,I struggle to fall asleep,I wake up a lot sooner than I would like','I struggle to fall asleep,I wake up a lot sooner than I would like',
                  'Good,I wake up a lot sooner than I would like']:
        return 0
    else:
        return 1
        
df['How do you sleep at night?'] = df['How do you sleep at night?'].apply(categorize_sleep)

df['How do you sleep at night?'].unique()

array([1, 0])

In [32]:
df['How many times per week do you eat fastfood?'].unique()

array([0, 1, 3, 2, 5, 4, 6, 7])

In [33]:
for column in questions_chi_square:
    print(column, df[column].unique())


Nuts ['Yes' 'Sometimes' 'No' "Don't know"]
How stress affects on your eating habits? ["I'm starting to eat more" "I'm starting to eat less"
 "Doesn't have any effect" 'I have no stress']
Legumes ['Yes' 'No' 'Sometimes' "Don't know"]
Cow's milk ['Yes' 'No' 'Sometimes' "Don't know"]
Do you eat alone or togheter? ['Often Together' 'Often Alone' 'Often Alone,Alone at dinner'
 'Often Together,Alone at breakfast,Alone at lunch'
 'Often Together,Alone at breakfast' 'Often Together,Often Alone'
 'Often Together,Alone at lunch'
 'Often Alone,Alone at breakfast,Alone at dinner'
 'Often Together,Alone at lunch,Alone at dinner'
 'Often Alone,Alone at lunch,Alone at dinner' 'Alone at breakfast'
 'Often Alone,Alone at breakfast'
 'Often Together,Often Alone,Alone at breakfast'
 'Alone at breakfast,Alone at dinner'
 'Often Together,Alone at breakfast,Alone at dinner'
 'Alone at lunch,Alone at dinner'
 'Alone at breakfast,Alone at lunch,Alone at dinner' 'Alone at dinner'
 'Often Together,Often Alone,A

In [34]:
# меняем "не знаю" на "нет", если ответов не знаю очень мало
for column in questions_chi_square:
    if "Don't know" in df[column].unique():
        count_dkn = df[column].value_counts()["Don't know"]
        if count_dkn < 100:
            df[column] = df[column].replace('NO', 'No').replace("Don't know", 'No')
            print(column,count_dkn)

Nuts 16
Legumes 5
Cow's milk 1
Raw vegetables 2
Low-fat white yogurt 15
Fresh cheeses 4
Tofu 97
Whole grains food 2
Eggs 56
Vegetable drinks (eg soy milk) 14
Processed Meat (es prosciutto) 57


In [35]:
for column in questions_chi_square:
    print(f'{column}\n')
    contingency_table = pd.crosstab(df[column], df['Average number of steps per day?'])
    chi2, p_value, dof, expected = chi2_contingency(contingency_table)
    print(f'Chi-squared: {chi2}')
    print(f'p-value: {p_value}')
    print("Expected:\n")
    print(expected)
    print("\nContingency table:\n")
    print(contingency_table)
    print('\n=======================================\n')

Nuts

Chi-squared: 28.013548800141724
p-value: 0.00047167918237104613
Expected:

[[ 37.33842239  74.24766751  79.39779474  42.91772689  19.09838846]
 [ 61.68956743 122.67005937 131.17896522  70.90754877  31.5538592 ]
 [ 74.97201018 149.08227311 159.42324003  86.17472434  38.34775233]]

Contingency table:

Average number of steps per day?   0    1    2   3   4
Nuts                                                  
No                                37   73   58  56  29
Sometimes                         64  134  120  70  30
Yes                               73  139  192  74  30


How stress affects on your eating habits?

Chi-squared: 44.61358410029055
p-value: 1.1999901357530443e-05
Expected:

[[ 37.33842239  74.24766751  79.39779474  42.91772689  19.09838846]
 [ 14.75826972  29.34690416  31.38252757  16.96352841   7.54877014]
 [ 59.03307888 117.38761662 125.53011026  67.85411366  30.19508058]
 [ 62.87022901 125.0178117  133.68956743  72.26463104  32.15776081]]

Contingency table:

Avera

### Выводы:

1. How stress affects on your eating habits?
   - Хи-квадрат: 44.61
   - p-value: 1.20e-05 < 0.05
   - Таблицы указывают на существенную связь между воздействием стресса на пищевые привычки
   
    Вывод: Присутствие статистически значимой связи, то есть воздействие стресса оказывает влияние на кол-во шагов

2. Eggs
   - Критерий Хи-квадрат: 39.29
   - Значение p-value: 4.35e-06 < 0.05
   
   Вывод: Результаты указывают на наличие статистической связи между потреблением яиц и кол-вом шагов


3. Do you wake up to eat at night?
   - Критерий Хи-квадрат: 67.02
   - Значение p-value: 1.15e-09 < 0.05
   
   Вывод: существует высокая статистическая связь между привычкой просыпаться ночью, чтобы поесть, и кол-вом шагов
  
4. Fresh cheeses

    - Критерий Хи-квадрат: 53.77027394797432
    - Значение p-value: 7.649141622731944e-09 < 0.05

    Выводы: cуществует статистически значимая связь между употреблением свежих сыров и количеством пройденных шагов в день активность.

5. Processed Meat (es prosciutto)

    - Критерий Хи-квадрат: 17.339540694216094
    - Значение p-value: 0.02676138948798746 < 0.05

    Выводы: есть статистически значимые связи между его потреблением и количеством шагов в день

6. Fish

    - Критерий Хи-квадрат: 10.61775393946618
    - Значение p-value: 0.22431221994528672 > 0.05

    Выводы:для употребления рыбы наблюдается отсутствие статистически значимой связи с количеством пройденных шагов в день

7. When are you hungry during the day?

    - Критерий Хи-квадрат: 71.98386541606149
    - Значение p-value: 1.0903172892657588e-06 < 0.05

    Выводы: высокая статистическая связь между временем дня, когда человек чувствует голод, и количеством пройденных шагов в день
    
8. Vegetable drinks (eg soy milk)

    - Критерий Хи-квадрат: 24.954086195975343
    - Значение p-value: 0.001582651251801338 < 0.05

   Выводы: употребление овощных напитков несет статистически значимую связь с количеством пройденных шагов в день

9. Do you eat alone or together?

    - Критерий Хи-квадрат: 128.5714938063711
    - Значение p-value: 0.051459738739660754 ~ 0.05

    Выводы: статистическая связь между тем, ест ли человек в одиночку или вместе с другими и количеством пройденных шагов в день

10. Do you play a sport (at least 5 hours/week)?

    - Критерий Хи-квадрат: 44.290406605712825
    - Значение p-value: 5.583691363832958e-09 < 0.05

    Выводы:  существует значимая связь между занятием спортом (хотя бы 5 часов в неделю) и количеством пройденных шагов в день
11. Do you ever eat distracted or not at the table?

    - Критерий Хи-квадрат: 13.350389694172733
    - Значение p-value: 0.009684765775275214 < 0.05

    Выводы: отвлеченное питание или прием пищи не за столом имеют статистически значимую связь и количеством пройденных шагов в день

12. Do you happen to eat uncontrollably even if you're not hungry?

    - Критерий Хи-квадрат: 14.486595198312926
    - Значение p-value: 0.2707196902422115 > 0.05

    Выводы: отсутствует статистически значимая связь между перекусами ночью и количеством пройденных шагов в день на основе значения p-value

13. Raw vegetables

    - Критерий Хи-квадрат: 57.370612958401786
    - Значение p-value: 1.5244462916165813e-09 < 0.05

    Выводы: потребление сырых овощей имеет статистически значимую связь с количеством пройденных шагов в день
14. Cow's milk

    - Критерий Хи-квадрат: 20.366521832722
    - Значение p-value: 0.009034846789339974 < 0.05

    Выводы: потребление коровьего молока может влиять на количество пройденных шагов в день

15. Meat

    - Критерий Хи-квадрат: 60.18680485174055
    - Значение p-value: 4.2838003291894864e-10 < 0.05

    Выводы: потребление мяса имеет статистически значимую связь с количеством пройденных шагов в день
16. Nuts

    - Критерий Хи-квадрат: 28.013548800141724
    - Значение p-value: 0.00047167918237104613 < 0.05

    Выводы: потребление орехов имеет статистически значимую связь с количеством пройденных шагов в день

17. Tofu

    - Критерий Хи-квадрат: 20.760093914731808
    - Значение p-value: 0.007812771593479565 < 0.05

    Выводы: потребление тофу может иметь влияние на количество пройденных шагов в день

18. What kind of snacks do you prefer?

    - Критерий Хи-квадрат: 161.25782898450433 - big
    - Значение p-value: 0.10553513979651609 > 0.05

    Выводы: существует связь между предпочтениями в закусках и количеством пройденных шагов в день
19. Legumes

    - Критерий Хи-квадрат: 38.86799274250605
    - Значение p-value: 5.200457196677014e-06  < 0.05

    Выводы: употребление бобов имеет статистически значимую связь с количеством пройденных шагов в день

20. How many hours a day do you sleep?

    - Критерий Хи-квадрат: 58.3473642324624
    - Значение p-value: 9.911966705189435e-07 < 0.05

    Выводы: статистически значимая связь между количеством часов сна в день и количеством пройденных шагов в день

21. SEX

    - Критерий Хи-квадрат: 14.882946302799656
    - Значение p-value: 0.004950254379033155 < 0.05

    Выводы: пол человека имеет статистически значимую связь с количеством пройденных шагов в день

22. Do you eat differently at the weekend?

    - Критерий Хи-квадрат: 57.18562064286119
    - Значение p-value: 7.317090447971519e-08 < 0.05

    Выводы: статистически значимая связь между тем, ест ли человек по-разному в выходные дни, и количеством пройденных шагов в день

23. How do you sleep at night?

    - Критерий Хи-квадрат: 13.849189347082683
    - Значение p-value: 0.007792266586501632 < 0.05

    Выводы: статистически значимая связь между способом сна ночью и количеством пройденных шагов в день
24. Do you ever miss meals?

    - Критерий Хи-квадрат: 103.30267276973957
    - Значение p-value: 3.0098697559056974e-05 < 0.05

    Выводы: иногда пропускаемые приемы пищи имеют статистически значимую связь с количеством пройденных шагов в день
25. Do you eat fast?

    - Критерий Хи-квадрат: 26.78383872994287
    - Значение p-value: 2.198275070276208e-05 < 0.05

    Выводы: быстрое питание имеет статистически значимую связь с количеством пройденных шагов в день